### ISSUE TITLE
FlakeIds referring to FlakeIds

### ISSUE THREAD

**Issue by CSVdB** [tweag/nix#37]
Right now, registries are maps from FlakeIds to FlakeRefs which can't be FlakeIds.
However, there are use cases where one would like a FlakeId to refer to a FlakeId.

Example:
Imagine the new release nixpkgs-19.03 just came out and I'm writing a package which depends on that. I now want to check whether my package also works with nixpkgs-18.09. What I want to do is
```nix build --flake nixpkgs nixpkgs-18.09```
and in my registry have
```
{
    nixpkgs: nixpkgs-19.03;
    nikpkgs-19.03: github:NixOS/nixpkgs/nixpkgs-19.03;
    nikpkgs-18.09: github:NixOS/nixpkgs/nixpkgs-18.09;
    ...
}
```

Implementing this would only require the "isDirect" check to be removed from `lookupFlake`.
However, this allows for circular lookups, which need to be avoided somehow.

**Comment by edolstra** [comment]
BTW, the way I see it, we wouldn't have entries like `nixpkgs-18.09` in the global registry. Instead, you would use a flake ID + a ref override, i.e. `nixpkgs/18.09`. So if the registry maps the flake ID `nixpkgs` to `github:NixOS/nixpkgs/19.03`, then `nixpkgs/18.09` resolves to `github:NixOS/nixpkgs/18.09`.

**Comment by CSVdB** [comment]
I have a few problems with that though:
1. I might want to refer to different base-flakerefs using `nixpkgs-18.03` and `nixpkgs-18.09` (for example, with the latest version I might still be on the official release, but on the old version I might be on my own fork).
2. The `ref` and `rev` of a FlakeId are currently entirely ignored. FlakeIds are resolved in `lookupFlake`, where only the the `id` itself is used.
